File: 0804-unique-morse-code-words/0804-unique-morse-code-words.ts
const MORSE: string[] = [
    ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---",
    "-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-",
    "...-", ".--", "-..-", "-.--", "--.."
];

function uniqueMorseRepresentations(words: string[]): number {
    const hashMap: { [key: string]: boolean } = {};
    for (const word of words) {
        const morseRepresentation = getMorseRepresentation(word);
        hashMap[morseRepresentation] = true;
    }
    return Object.keys(hashMap).length;
}

function getMorseRepresentation(word: string): string {
    let answer = "";
    for (const char of word) {
        const asciiValue = charToAscii(char);
        answer += MORSE[asciiValue];
    }
    return answer;
}

function charToAscii(char: string): number {
    return char.charCodeAt(0) - 'a'.charCodeAt(0);
}

File: 0804-unique-morse-code-words/0804-unique-morse-code-words.go
var morse []string = []string{".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..", "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-", "-.--", "--.."}

func uniqueMorseRepresentations(words []string) int {
	hashMap := make(map[string]bool)
	for _, v := range words {
		morseRepresentation := getMorseRepresentation(v)
		hashMap[morseRepresentation] = true
	}
	return len(hashMap)
}

func getMorseRepresentation(word string) string {
	var answer string
	for _, v := range word {
		asciiValue := charToAscii(v)
		answer += morse[asciiValue]
	}
	return answer
}

func charToAscii(char rune) int {
	return int(char - 'a')
}
